File: ruby/conformance/protocol_runner.rb
# frozen_string_literal: true

# Canonical mpp-tools protocol-conformance runner for the Ruby SDK.
#
# Speaks the per-language adapter ABI documented in
# `harness/src/protocol/README.md`: read one JSON request on stdin
#
#   { "op": "<operation>", "input": <op-specific> }
#
# and write one JSON response on stdout, then exit
#
#   { "success": true,  "result": <op-specific> }
#   { "success": false, "error": "<msg>", "error_type": "<type>" }
#
# Each op maps to the existing `solana-pay-kit` Ruby protocol functions (no
# conformance-only reimplementation):
#
#   challenge.parse  -> Protocol::Core::Headers.parse_www_authenticate
#   challenge.format -> Protocol::Core::Headers.format_www_authenticate
#   credential.parse -> Protocol::Core::Credential.from_authorization_header
#   credential.format-> Protocol::Core::Credential#to_authorization_header
#   receipt.parse    -> Protocol::Core::Headers.parse_receipt
#   receipt.format   -> Protocol::Core::Headers.format_receipt
#   base64url.encode -> PayCore::Base64Url.encode
#   base64url.decode -> PayCore::Base64Url.decode
#   challenge.id     -> Protocol::Core::Challenge.compute_id (over JCS+base64url request)
#
# The `error_type` vocabulary mirrors the TypeScript reference runner:
# parse_error (.parse), format_error (.format), encoding_error (base64url),
# generation_error (challenge.id).

require "json"

lib = File.expand_path("../lib", __dir__)
$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)

# The protocol/core files declare `module PayKit::Protocols::Mpp` and assume
# the parent namespaces already exist; predeclare them so the core codec files
# can be required in isolation without dragging in the full server stack.
module PayKit
  module Protocols
    module Mpp
    end
  end
end

require "pay_core/base64_url"
require "pay_core/json"
require "pay_kit/protocols/mpp/protocol/core/challenge"
require "pay_kit/protocols/mpp/protocol/core/credential"
require "pay_kit/protocols/mpp/protocol/core/receipt"
require "pay_kit/protocols/mpp/protocol/core/headers"

module ConformanceRunner
  Core = PayKit::Protocols::Mpp::Protocol::Core
  Headers = Core::Headers

  module_function

  def ok(result)
    {success: true, result: result}
  end

  def fail(message, error_type)
    # SDK error messages may embed raw bytes from an invalid base64url payload
    # (e.g. a request param that decodes to non-UTF-8). Scrub to valid UTF-8 so
    # the ABI response can be JSON-serialized instead of crashing the runner.
    scrubbed = message.to_s.encode("UTF-8", invalid: :replace, undef: :replace)
    {success: false, error: scrubbed, error_type: error_type}
  end

  def header_of(input)
    input.fetch("header")
  end

  def text_of(input)
    input.fetch("text")
  end

  # challenge.parse golden shape carries `request` as the DECODED JSON object,
  # so expose the parsed Challenge with its request decoded back to an object.
  def challenge_to_object(challenge)
    obj = {
      "id" => challenge.id,
      "realm" => challenge.realm,
      "method" => challenge.method,
      "intent" => challenge.intent,
      "request" => challenge.decode_request
    }
    obj["expires"] = challenge.expires unless challenge.expires.nil?
    obj["description"] = challenge.description unless challenge.description.nil?
    obj["digest"] = challenge.digest unless challenge.digest.nil?
    obj["opaque"] = challenge.opaque unless challenge.opaque.nil?
    obj
  end

  # challenge.format input is the canonical challenge object (request is a
  # decoded JSON object); rebuild a Challenge by re-encoding request to the
  # base64url(JCS) wire form the codec expects, then format the header.
  def challenge_from_object(input)
    encoded_request = ::PayCore::Base64Url.encode(
      ::PayCore::Json.canonical_generate(input.fetch("request"))
    )
    Core::Challenge.new(
      id: input.fetch("id"),
      realm: input.fetch("realm"),
      method: input.fetch("method"),
      intent: input.fetch("intent"),
      request: encoded_request,
      expires: input["expires"],
      description: input["description"],
      digest: input["digest"],
      opaque: input["opaque"]
    )
  end

  # credential.format input is the canonical credential object whose nested
  # `challenge.request` is a DECODED JSON object. Re-encode it to the
  # base64url(JCS) wire string the ChallengeEcho echoes verbatim.
  def credential_from_object(input)
    challenge = input.fetch("challenge").dup
    if challenge["request"].is_a?(Hash) || challenge["request"].is_a?(Array)
      challenge["request"] = ::PayCore::Base64Url.encode(
        ::PayCore::Json.canonical_generate(challenge["request"])
      )
    end
    echo = Core::ChallengeEcho.from_h(challenge)
    Core::Credential.new(
      challenge: echo,
      payload: input.fetch("payload"),
      source: input["source"]
    )
  end

  def credential_to_object(credential)
    credential.to_h
  end

  def receipt_from_object(input)
    Core::Receipt.new(
      status: input.fetch("status"),
      method: input.fetch("method"),
      reference: input.fetch("reference"),
      challenge_id: input["challengeId"],
      external_id: input["externalId"],
      timestamp: input.fetch("timestamp")
    )
  end

  # challenge.id input passes `request` as a decoded object and `opaque` as an
  # already-serialized string. Canonicalize+base64url the request, then run the
  # HMAC over the canonical pipe layout (Challenge.compute_id).
  def challenge_id(input)
    encoded_request = ::PayCore::Base64Url.encode(
      ::PayCore::Json.canonical_generate(input.fetch("request", {}))
    )
    Core::Challenge.compute_id(
      secret_key: input.fetch("secretKey"),
      realm: input.fetch("realm", ""),
      method: input.fetch("method", ""),
      intent: input.fetch("intent", ""),
      request: encoded_request,
      expires: input["expires"],
      digest: input["digest"],
      opaque: input["opaque"]
    )
  end

  def dispatch(op, input)
    case op
    when "challenge.parse"
      ok(challenge_to_object(Headers.parse_www_authenticate(header_of(input))))
    when "challenge.format"
      ok({"header" => Headers.format_www_authenticate(challenge_from_object(input))})
    when "credential.parse"
      ok(credential_to_object(Core::Credential.from_authorization_header(header_of(input))))
    when "credential.format"
      ok({"header" => credential_from_object(input).to_authorization_header})
    when "receipt.parse"
      ok(Headers.parse_receipt(header_of(input)).to_h)
    when "receipt.format"
      ok({"header" => Headers.format_receipt(receipt_from_object(input))})
    when "base64url.encode"
      # base64url.encode result is the encoded text.
      ok({"text" => ::PayCore::Base64Url.encode(text_of(input))})
    when "base64url.decode"
      # base64url.decode yields UTF-8 text; the SDK returns binary bytes.
      decoded = ::PayCore::Base64Url.decode(text_of(input))
      ok({"text" => decoded.force_encoding(Encoding::UTF_8)})
    when "challenge.id"
      ok({"id" => challenge_id(input)})
    else
      fail("Unknown operation: #{op}", "unsupported_operation")
    end
  rescue KeyError, ArgumentError, TypeError, JSON::ParserError => error
    fail(error.message, error_type_for(op))
  end

  def error_type_for(op)
    return "parse_error" if op.end_with?(".parse")
    return "format_error" if op.end_with?(".format")
    return "encoding_error" if op.start_with?("base64url.")
    return "generation_error" if op == "challenge.id"

    "unknown_error"
  end

  def run(raw)
    request = JSON.parse(raw)
    dispatch(request.fetch("op"), request.fetch("input"))
  rescue JSON::ParserError => error
    fail("invalid request JSON: #{error.message}", "unknown_error")
  end
end

if $PROGRAM_NAME == __FILE__
  raw = $stdin.read.to_s.strip
  response = ConformanceRunner.run(raw)
  $stdout.write(JSON.generate(response))
end

File: php/conformance/protocol-runner.php
<?php

declare(strict_types=1);

/**
 * PHP mpp-protocol primitive conformance runner.
 *
 * Speaks the canonical mpp-tools adapter ABI used by the protocol-conformance
 * layer (harness/src/protocol/runners/typescript.ts is the reference): read one
 *   { "op": "<operation>", "input": <op-specific> }
 * request as JSON on stdin and write one
 *   { "success": true,  "result": <op-specific> }
 *   { "success": false, "error": "<msg>", "error_type": "<type>" }
 * response as JSON on stdout. This is the protocol-primitive counterpart to the
 * charge/x402 conformance runner (conformance/runner.php); the two ABIs are
 * distinct, so they live in separate entrypoints.
 *
 * The operations map onto the PHP SDK's existing protocol surface
 * (PayKit\Protocols\Mpp\Core\*, PayKit\PayCore\Wire\*) per the Reference map:
 *
 *   challenge.parse   -> Headers::parseWwwAuthenticate
 *   challenge.format  -> Headers::formatWwwAuthenticate
 *   credential.parse  -> Credential::fromAuthorizationHeader
 *   credential.format -> Credential::toAuthorizationHeader
 *   receipt.parse     -> Headers::parseReceipt
 *   receipt.format    -> Headers::formatReceipt
 *   base64url.encode  -> Base64Url::encode
 *   base64url.decode  -> Base64Url::decode
 *   challenge.id      -> Challenge::computeId
 *
 * Every operation here is implemented by the PHP SDK, so this runner never
 * emits unsupported_operation for the canonical op set; it only does so for an
 * op string the SDK has no surface for (forward-compat guard).
 *
 * RPC-free and deterministic: no validator, no network, no signing. Pure
 * wire codec + HMAC math.
 */

error_reporting(error_reporting() & ~E_DEPRECATED & ~E_USER_DEPRECATED);
ini_set('display_errors', 'stderr');

require __DIR__ . '/../vendor/autoload.php';

use PayKit\PayCore\Wire\Base64Url;
use PayKit\PayCore\Wire\Json;
use PayKit\Protocols\Mpp\Core\Challenge;
use PayKit\Protocols\Mpp\Core\Credential;
use PayKit\Protocols\Mpp\Core\Headers;

/**
 * Re-encode a JSON `request` slot to the SDK's stored base64url(JCS) form,
 * preserving the object/array distinction that PHP's associative decode
 * erases.
 *
 * PHP collapses both `{}` and `[]` to the same empty `array`, and
 * Json::canonicalize emits a list `[]` for it, so a request that is (or nests)
 * an empty JSON object would canonicalize to `[]` (`W10`) instead of `{}`
 * (`e30`). The canonical wire and every other SDK encode an empty object as
 * `{}`. This re-decodes the slot from the ORIGINAL JSON text with object
 * preservation (json_decode without associative) and recurses, deferring every
 * scalar and every non-empty container to the SDK canonicalizer
 * (Json::canonicalize) — only the empty-object ambiguity is resolved here, not
 * a second JCS implementation.
 *
 * `$requestJson` is the raw JSON text of the request slot; `$fallback` is the
 * already-decoded associative value used when no raw text is available.
 */
function request_to_base64url(?string $requestJson, mixed $fallback): string
{
    if ($requestJson !== null) {
        $preserving = json_decode($requestJson, false, flags: JSON_THROW_ON_ERROR);
        return Base64Url::encode(canonicalize_preserving_objects($preserving));
    }
    if (is_array($fallback) && $fallback !== []) {
        return Base64Url::encodeJson($fallback);
    }
    // Absent or empty object -> canonical `{}`.
    return Base64Url::encode('{}');
}

/**
 * RFC 8785 canonical JSON that keeps empty objects as `{}`.
 *
 * A value with no empty-object descendants is canonicalized in one shot by the
 * SDK (Json::canonicalize on its associative form), so the JCS math (key order,
 * number/string escaping) is always the SDK's. Only when an empty object is
 * present (which PHP's array type cannot distinguish from `[]`) does this
 * recurse to splice the literal `{}` token into the SDK-canonical object,
 * preserving the SDK's key ordering by canonicalizing the object with that key
 * temporarily mapped to a unique sentinel and substituting the `{}` back in.
 */
function canonicalize_preserving_objects(mixed $value): string
{
    if (!has_empty_object($value)) {
        return Json::canonicalize(object_tree_to_assoc($value));
    }
    if ($value instanceof \stdClass) {
        $entries = get_object_vars($value);
        if ($entries === []) {
            return '{}';
        }
        // Map every value to a unique scalar sentinel string, canonicalize the
        // object through the SDK (exact JCS key order + key escaping), then
        // replace each quoted sentinel with its child's canonical bytes,
        // recursing so nested empty objects resolve to `{}`.
        $sentinels = [];
        $substitutions = [];
        $i = 0;
        foreach ($entries as $k => $v) {
            $token = "\u{0000}sentinel{$i}\u{0000}";
            $sentinels[(string) $k] = $token;
            // The SDK emits the sentinel as a JSON string literal; map that
            // exact literal to the child's already-canonical bytes.
            $substitutions[Json::canonicalize($token)] = canonicalize_preserving_objects($v);
            $i++;
        }
        return strtr(Json::canonicalize($sentinels), $substitutions);
    }
    if (is_array($value)) {
        $parts = array_map('canonicalize_preserving_objects', $value);
        return '[' . implode(',', $parts) . ']';
    }

    return Json::canonicalize($value);
}

/**
 * Recursively convert a json_decode object tree (stdClass + array) to the
 * associative-array shape the SDK canonicalizer consumes. Only used for values
 * with no empty objects, so the lossy `{}` -> `[]` collapse never bites.
 */
function object_tree_to_assoc(mixed $value): mixed
{
    if ($value instanceof \stdClass) {
        $out = [];
        foreach (get_object_vars($value) as $k => $v) {
            $out[(string) $k] = object_tree_to_assoc($v);
        }
        return $out;
    }
    if (is_array($value)) {
        return array_map('object_tree_to_assoc', $value);
    }
    return $value;
}

/**
 * True when $value is, or contains anywhere, an empty JSON object (`{}`),
 * decoded with object preservation as an empty stdClass.
 */
function has_empty_object(mixed $value): bool
{
    if ($value instanceof \stdClass) {
        $entries = get_object_vars($value);
        if ($entries === []) {
            return true;
        }
        foreach ($entries as $v) {
            if (has_empty_object($v)) {
                return true;
            }
        }
        return false;
    }
    if (is_array($value)) {
        foreach ($value as $v) {
            if (has_empty_object($v)) {
                return true;
            }
        }
    }
    return false;
}

/**
 * Read the entire request JSON from stdin.
 */
function read_stdin(): string
{
    $raw = stream_get_contents(STDIN);
    if (!is_string($raw)) {
        throw new RuntimeException('php protocol runner failed to read stdin');
    }
    $trimmed = trim($raw);
    if ($trimmed === '') {
        throw new RuntimeException('php protocol runner received empty stdin');
    }
    return $trimmed;
}

/**
 * @param array<string, mixed> $result
 */
function emit(array $result): void
{
    fwrite(STDOUT, json_encode($result, JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE) . "\n");
}

/**
 * Map the PHP SDK's parsed Challenge onto the canonical adapter-ABI challenge
 * object shape. The canonical object carries `request` as a DECODED JSON
 * object (not the base64url string the PHP value type stores), so we decode it
 * here, exactly as mppx's deserialized challenge surfaces it.
 *
 * The PHP parser intentionally drops unknown auth-params (including
 * `description`), mirroring the rust spine's strict PaymentChallenge struct.
 * The canonical golden object echoes a top-level `description` when the wire
 * carried one, so scenarios that pin a description diverge structurally on
 * that single field; the runner emits the SDK's true output and the driver
 * reports the gap honestly.
 *
 * @return array<string, mixed>
 */
function challenge_to_canonical(Challenge $challenge): array
{
    $object = [
        'id' => $challenge->id,
        'realm' => $challenge->realm,
        'method' => $challenge->method,
        'intent' => $challenge->intent,
        // Decode the stored base64url(JCS) request with OBJECT preservation so an
        // empty `{}` survives as a JSON object, not the `[]` PHP's associative
        // decode + json_encode would otherwise emit. The canonical wire and
        // every other SDK echo an empty request as `{}`.
        'request' => json_decode(Base64Url::decode($challenge->request), false, flags: JSON_THROW_ON_ERROR),
    ];
    if ($challenge->description !== null) {
        $object['description'] = $challenge->description;
    }
    if ($challenge->expires !== '') {
        $object['expires'] = $challenge->expires;
    }
    if ($challenge->digest !== '') {
        $object['digest'] = $challenge->digest;
    }
    if ($challenge->opaque !== null) {
        $object['opaque'] = $challenge->opaque;
    }

    return $object;
}

/**
 * Build a PHP Challenge value type from the canonical adapter-ABI challenge
 * object. `request` arrives as a decoded JSON object and must be re-encoded to
 * the base64url(JCS) form the PHP value type stores. A non-lowercase or empty
 * required field surfaces through the Challenge constructor as a format_error.
 *
 * @param array<string, mixed> $object
 */
function challenge_from_canonical(array $object, mixed $requestObj = null): Challenge
{
    $request = $object['request'] ?? [];
    if (!is_array($request)) {
        throw new InvalidArgumentException('challenge.request must be an object');
    }

    // Encode the request from the object-preserving input so an empty `{}`
    // canonicalizes to `e30`, not the `[]` -> `W10` PHP's associative array
    // would yield. The JCS math stays the SDK canonicalizer's.
    $encodedRequest = $requestObj !== null
        ? Base64Url::encode(canonicalize_preserving_objects($requestObj))
        : Base64Url::encodeJson($request);

    return new Challenge(
        id: as_string($object['id'] ?? null, 'id'),
        realm: as_string($object['realm'] ?? null, 'realm'),
        method: as_string($object['method'] ?? null, 'method'),
        intent: as_string($object['intent'] ?? null, 'intent'),
        request: $encodedRequest,
        expires: opt_string($object['expires'] ?? null),
        digest: opt_string($object['digest'] ?? null),
        opaque: isset($object['opaque']) ? as_string($object['opaque'], 'opaque') : null,
        description: isset($object['description']) ? as_string($object['description'], 'description') : null,
    );
}

/**
 * Build a PHP Credential from the canonical adapter-ABI credential object.
 * The canonical object nests the challenge with `request` as a decoded object;
 * ChallengeEcho::fromArray already re-encodes a nested object request to the
 * stored base64url(JCS) string, so the object is consumed as-is.
 *
 * @param array<string, mixed> $object
 */
function credential_from_canonical(array $object): Credential
{
    // Round-trip through the SDK's own credential decoder by serializing the
    // canonical object to the SDK's Authorization wire and parsing it back.
    // This drives Credential::fromAuthorizationHeader (the parse surface) for
    // the format direction, so format and parse share one code path and the
    // SDK owns every field-shape decision (challenge echo, payload, source).
    $header = 'Payment ' . Base64Url::encodeJson($object);
    return Credential::fromAuthorizationHeader($header);
}

function as_string(mixed $value, string $field): string
{
    if (!is_string($value)) {
        throw new InvalidArgumentException(sprintf('field "%s" must be a string', $field));
    }
    return $value;
}

function opt_string(mixed $value): string
{
    return is_string($value) ? $value : '';
}

/**
 * @param array<string, mixed> $input
 */
function require_header(array $input): string
{
    $header = $input['header'] ?? null;
    if (!is_string($header)) {
        throw new InvalidArgumentException('input.header must be a string');
    }
    return $header;
}

/**
 * @param array<string, mixed> $input
 */
function require_text(array $input): string
{
    $text = $input['text'] ?? null;
    if (!is_string($text)) {
        throw new InvalidArgumentException('input.text must be a string');
    }
    return $text;
}

/**
 * Dispatch a single adapter-ABI request to the PHP SDK protocol surface.
 *
 * `$input` is the associative-decoded input; `$inputObj` is the same input
 * decoded with object preservation (stdClass for objects), used only to keep
 * the empty-object `{}` distinction PHP's associative decode erases — needed
 * for the byte-exact challenge.id HMAC.
 *
 * @param array<string, mixed> $input
 * @return array<string, mixed> the success result payload (op-specific)
 */
function dispatch(string $op, array $input, mixed $inputObj): array
{
    switch ($op) {
        case 'challenge.parse':
            return challenge_to_canonical(Headers::parseWwwAuthenticate(require_header($input)));

        case 'challenge.format':
            $requestObj = ($inputObj instanceof \stdClass && property_exists($inputObj, 'request'))
                ? $inputObj->request
                : null;
            return ['header' => Headers::formatWwwAuthenticate(challenge_from_canonical($input, $requestObj))];

        case 'credential.parse':
            return Credential::fromAuthorizationHeader(require_header($input))->toArray();

        case 'credential.format':
            return ['header' => credential_from_canonical($input)->toAuthorizationHeader()];

        case 'receipt.parse':
            return Headers::parseReceipt(require_header($input))->toArray();

        case 'receipt.format':
            // Re-issue the SDK's own receipt parse over a wire built from the
            // canonical object, so the format direction drives the SDK
            // formatter (Headers::formatReceipt) on a faithfully-typed Receipt.
            $receipt = Headers::parseReceipt(Base64Url::encodeJson($input));
            return ['header' => Headers::formatReceipt($receipt)];

        case 'base64url.encode':
            return ['text' => Base64Url::encode(require_text($input))];

        case 'base64url.decode':
            return ['text' => Base64Url::decode(require_text($input))];

        case 'challenge.id':
            $request = $input['request'] ?? [];
            if (!is_array($request)) {
                throw new InvalidArgumentException('challenge.id input.request must be an object');
            }
            // Encode the request from the object-preserving input so an empty
            // `{}` canonicalizes to `e30`, not `[]` -> `W10`. The HMAC math,
            // pipe layout, and base64url all stay the SDK's (Challenge::computeId).
            $requestObj = ($inputObj instanceof \stdClass && property_exists($inputObj, 'request'))
                ? $inputObj->request
                : $request;
            $id = Challenge::computeId(
                as_string($input['secretKey'] ?? null, 'secretKey'),
                opt_string($input['realm'] ?? null),
                opt_string($input['method'] ?? null),
                opt_string($input['intent'] ?? null),
                request_to_base64url(json_encode($requestObj, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE) ?: null, $request),
                opt_string($input['expires'] ?? null),
                opt_string($input['digest'] ?? null),
                isset($input['opaque']) ? as_string($input['opaque'], 'opaque') : null,
            );
            return ['id' => $id];

        default:
            throw new RuntimeException('UNSUPPORTED_OPERATION');
    }
}

/**
 * Classify a thrown error into the canonical error_type vocabulary keyed on
 * the operation suffix (mirrors the TS reference runner's catch arm).
 */
function error_type_for(string $op): string
{
    if (str_ends_with($op, '.parse')) {
        return 'parse_error';
    }
    if (str_ends_with($op, '.format')) {
        return 'format_error';
    }
    if (str_starts_with($op, 'base64url.')) {
        return 'encoding_error';
    }
    if ($op === 'challenge.id') {
        return 'generation_error';
    }
    return 'unknown_error';
}

try {
    $raw = read_stdin();
    $request = json_decode($raw, true, flags: JSON_THROW_ON_ERROR);
    if (!is_array($request)) {
        throw new RuntimeException('request must be a JSON object');
    }
    $op = $request['op'] ?? null;
    if (!is_string($op) || $op === '') {
        throw new RuntimeException('request.op must be a non-empty string');
    }
    $input = $request['input'] ?? [];
    if (!is_array($input)) {
        throw new RuntimeException('request.input must be an object');
    }
    // Object-preserving copy of input so the empty-object `{}` distinction
    // survives for challenge.id (PHP's associative decode collapses `{}` to []).
    $requestObj = json_decode($raw, false, flags: JSON_THROW_ON_ERROR);
    $inputObj = ($requestObj instanceof \stdClass && property_exists($requestObj, 'input'))
        ? $requestObj->input
        : null;

    try {
        $result = dispatch($op, $input, $inputObj);
        emit(['success' => true, 'result' => $result]);
    } catch (Throwable $error) {
        if ($error->getMessage() === 'UNSUPPORTED_OPERATION') {
            emit([
                'success' => false,
                'error' => 'php protocol runner does not implement operation: ' . $op,
                'error_type' => 'unsupported_operation',
            ]);
        } else {
            emit([
                'success' => false,
                'error' => $error->getMessage(),
                'error_type' => error_type_for($op),
            ]);
        }
    }
} catch (Throwable $fatal) {
    fwrite(STDERR, 'php protocol runner fatal: ' . $fatal->getMessage() . "\n");
    emit([
        'success' => false,
        'error' => $fatal->getMessage(),
        'error_type' => 'unknown_error',
    ]);
    exit(1);
}
